sdk/python/toolchain/uv.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package toolchain
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+)
+
+type uv struct {
+	// The absolute path to the virtual env.
+	virtualenvPath string
+	// The root directory of the project.
+	root string
+}
+
+var minUvVersion = semver.MustParse("0.4.26")
+
+var defaultVirtualEnv = ".venv"
```

Comment:
We default to `venv` for pip, but uv defaults to `.venv`, and I feel that `.venv` is also more common in Python-land. The docs also mention `.venv`:

> A common directory location for a virtual environment is .venv. 

https://docs.python.org/3/tutorial/venv.html#creating-virtual-environments

Do we mind the inconsistency? I am tempted to start using `.venv` for new pip projects (and update our templates to use that, if they do specify the `virtualenv` option.

Reply:
@julienp, thanks for your contribution. It's a good start. My 20 cents about the virtual environment folder is not an issue. We can define it in Pulumi.yaml:
```
name: pulumi-uv
description: A Python program to deploy a Kubernetes cluster on Google Cloud
runtime:
  name: python
  options:
    toolchain: poetry
    typechecker: pyright
    virtualenv: .venv
config:
  pulumi:tags:
    value:
      pulumi:template: kubernetes-gcp-python
```

Reply:
@julienp It is also possible to create a default virtual environment folder using `uv` as:

`uv venv venv`

When `venv` is a folder.

Reply:
Personally I don't mind the inconsistency here.  If the default for uv is `.venv`, that's probably what we should stick to.  We could definitely update the folder for pip etc. but I'm not sure it's very important to be consistent between package managers here.